### PR TITLE
KTOR-927 Add git hook to check commit message

### PR DESCRIPTION
**Pull request by e5l**
Fixes [KTOR-927](https://youtrack.jetbrains.com/issue/KTOR-927)

### CODE REVIEW

File: githook/prepare-commit-msg [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+    if issue_name == "fixup!":
+        return 0
+
+    if issue_name == "WIP":
+        return 0
+
+    if issue_name.startswith("KTOR-") and is_integer(issue_name[5:]):
+        return 0
+
+    print(f"Commit name should starts with issue number(like KTOR-1234). Current prefix is '{issue_name}'")
```
Comment (cy6erGn0m) [inline githook/prepare-commit-msg:31]: should start_

File: githook/prepare-commit-msg [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+
+def check_commit_message() -> int:
+    commit_message_filename: str = sys.argv[1]
+    with open(commit_message_filename) as file:
+        first_line: str = file.readline().strip()
+
+    issue_name = first_line.split(" ")[0]
+
+    # Allow fixup commits
+    if issue_name == "fixup!":
```
Comment (cy6erGn0m) [inline githook/prepare-commit-msg:22]: What about my favourite ~ ? 🙂 Would be nice to allow it as well

File: githook/prepare-commit-msg [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+        return True
+    except ValueError as _:
+        pass
+
+    return False
+
+def check_commit_message() -> int:
+    commit_message_filename: str = sys.argv[1]
+    with open(commit_message_filename) as file:
+        first_line: str = file.readline().strip()
```
Comment (cy6erGn0m) [inline githook/prepare-commit-msg:17]: Do we need to strip? Shouldn't we prohibit leading spaces instead?

**Review by cy6erGn0m — CHANGES_REQUESTED** [review]

File: githook/prepare-commit-msg [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+def is_integer(value: str) -> bool:
+    try:
+        int(value)
+        return True
+    except ValueError as _:
+        pass
+
+    return False
+
+def check_commit_message() -> int:
```
Comment (Stexxe) [inline githook/prepare-commit-msg:14]: Bad function name. What exactly does it check?

File: githook/prepare-commit-msg [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+        return 0
+
+    if issue_name == "WIP":
+        return 0
+
+    if issue_name.startswith("KTOR-") and is_integer(issue_name[5:]):
+        return 0
+
+    print(f"Commit name should starts with issue number(like KTOR-1234). Current prefix is '{issue_name}'")
+    return 1
```
Comment (Stexxe) [inline githook/prepare-commit-msg:32]: It's not obvious what the return value means.

File: githook/prepare-commit-msg [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+    issue_name = first_line.split(" ")[0]
+
+    # Allow fixup commits
+    if issue_name == "fixup!":
+        return 0
+
+    if issue_name == "WIP":
+        return 0
+
+    if issue_name.startswith("KTOR-") and is_integer(issue_name[5:]):
```
Comment (Stexxe) [inline githook/prepare-commit-msg:28]: Isn't it easier to just match with a regular expression?

**Review by Stexxe — CHANGES_REQUESTED** [review]

File: githook/prepare-commit-msg (hunk 1 above)
Comment (e5l) [reply githook/prepare-commit-msg:31]: thanks, will fix

Comment (cy6erGn0m) [reply githook/prepare-commit-msg:31]: Just a catch: we could use a real ktor ticket describing commit message policy instead of 1234

**Comment by rsinukov** [comment]
Would be nice to also update the contribution guide.

File: githook/prepare-commit-msg (hunk 2 above)
Comment (e5l) [reply githook/prepare-commit-msg:22]: Sure :) will add

File: githook/prepare-commit-msg (hunk 4 above)
Comment (e5l) [reply githook/prepare-commit-msg:14]: It's Python. I used the default pep8 style guide

File: githook/prepare-commit-msg (hunk 5 above)
Comment (e5l) [reply githook/prepare-commit-msg:32]: It's an exit code. Let me introduce named constants

File: githook/prepare-commit-msg (hunk 6 above)
Comment (e5l) [reply githook/prepare-commit-msg:28]: Nope. I would avoid using regex in any case. Right now it's easy to read

File: githook/prepare-commit-msg (hunk 5 above)
Comment (Stexxe) [reply githook/prepare-commit-msg:32]: I know what it's. You should have a function that checks the presence of an issue in the commit message and returns boolean.

**Review by Stexxe — CHANGES_REQUESTED** [review]

File: githook/prepare-commit-msg (hunk 4 above)
Comment (e5l) [reply githook/prepare-commit-msg:17]: how would you name it?

File: githook/prepare-commit-msg (hunk 5 above)
Comment (e5l) [reply githook/prepare-commit-msg]: Yep, let me refactor that

File: githook/prepare-commit-msg (hunk 6 above)
Comment (e5l) [reply githook/prepare-commit-msg:35]: Cause regex not so good from a maintenance and readability perspective.
Could you tell me why 2 explicit checks should be merged in a single(which you also have to parse during reading)?

File: CONTRIBUTING.md [hunk 7]
```diff
@@ -9,13 +9,15 @@ if there is already an open ticket and open a new one if not yet opened.
 
 Contributions are made using Github [pull requests](https://help.github.com/en/articles/about-pull-requests). 
 
-1. [Create](https://github.com/ktorio/ktor/compare) a new PR, your PR should request to merge to the **master** branch.
-2. Ensure that the description is clear, refer to an existing ticket/bug if applicable.
-3. When contributing a new feature, provide motivation and use-cases describing why 
+1. Open(or find) corresponding [YT](https://youtrack.jetbrains.com/issues/KTOR) issue.
```
Comment (rsinukov) [inline CONTRIBUTING.md:12]: missing space after `Open`

File: CONTRIBUTING.md [hunk 8]
```diff
@@ -9,13 +9,15 @@ if there is already an open ticket and open a new one if not yet opened.
...
+2. Please make sure that commit messages start with issue number(like `KTOR-1001`). You can copy `githook/prepare-commit-msg` to `.git/hooks` to enable automatic commit message check.
```
Comment (rsinukov) [inline CONTRIBUTING.md:13]: missing space after `like`

File: githook/prepare-commit-msg [hunk 9]
```diff
@@ -0,0 +1,42 @@
+#!/usr/bin/env python3
+
+import sys
+
+ALLOW = 0
+RESTRICT = 0
```
Comment (rsinukov) [inline githook/prepare-commit-msg:6]: should it be a non zero exit code?

**Review by rsinukov — CHANGES_REQUESTED** [review]

File: githook/prepare-commit-msg (hunk 4 above)
Comment (cy6erGn0m) [reply githook/prepare-commit-msg:17]: I would name it check commit message 😆

File: CONTRIBUTING.md (hunk 7 above)
Comment (cy6erGn0m) [reply CONTRIBUTING.md:12]: Missing "the"?

File: githook/prepare-commit-msg [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+
+def is_integer(value: str) -> bool:
+    try:
+        int(value)
+        return True
+    except ValueError as _:
+        pass
```
Comment (Stexxe) [inline githook/prepare-commit-msg:13]: `return False` can be place here

File: githook/prepare-commit-msg (hunk 4 above)
Comment (Stexxe) [reply githook/prepare-commit-msg:17]: `is_message_contain_issue_id` or something similar

**Review by Stexxe — CHANGES_REQUESTED** [review]

File: githook/prepare-commit-msg [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+
+    # Allow fixup commits
+    if issue_name == "fixup!":
+        return ALLOW
+
+    # Allow WIP commits
+    if issue_name == "WIP":
+        return ALLOW
+
+    if issue_name.startswith("~"):
```
Comment (cy6erGn0m) [inline githook/prepare-commit-msg:32]: I am curious, doesn't python has something like `x in [fizup, wip, ~]`

**Review by cy6erGn0m — CHANGES_REQUESTED** [review]

**Review by rsinukov — APPROVED** [review]

File: githook/prepare-commit-msg [hunk 12]
```diff
@@ -0,0 +1,41 @@
...
+    print(f"Commit name should start with the issue number(like KTOR-1001). Current prefix is '{issue_name}'")
+    return False
+
+
+if __name__ == "__main__":
+    commit_message_filename: str = sys.argv[1]
+    with open(commit_message_filename) as file:
+        first_line: str = file.readline()
+
+    exit_code = ALLOW if is_correct_commit_message(first_line) else RESTRICT
```
Comment (Stexxe) [inline githook/prepare-commit-msg:40]: Those constants are used only in one place and obstruct reading the code

**Review by Stexxe — APPROVED** [review]

**Review by cy6erGn0m — APPROVED** [review]

**Pull request merged**